orderbook/src/api/get_trades.rs

```diff
@@ -1,9 +1,150 @@
+use super::H160Wrapper;
+use crate::api::convert_get_trades_error_to_reply;
 use crate::database::{Database, TradeFilter};
 use anyhow::Result;
 use futures::TryStreamExt;
+use model::order::OrderUid;
 use model::trade::Trade;
+use serde::Deserialize;
+use std::convert::Infallible;
+use std::sync::Arc;
+use warp::reply::{Json, WithStatus};
+use warp::{hyper::StatusCode, Filter, Rejection, Reply};
 
-#[allow(dead_code)]
-pub async fn get_trades(db: &Database, filter: &TradeFilter) -> Result<Vec<Trade>> {
-    Ok(db.trades(filter).try_collect::<Vec<_>>().await?)
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Query {
+    pub order_uid: Option<OrderUid>,
+    pub owner: Option<H160Wrapper>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum TradeFilterError {
+    InvalidFilter(String),
+}
+
+impl Query {
+    fn trade_filter(&self) -> TradeFilter {
+        let to_h160 = |option: Option<&H160Wrapper>| option.map(|wrapper| wrapper.0);
+        TradeFilter {
+            order_uid: self.order_uid,
+            owner: to_h160(self.owner.as_ref()),
+        }
+    }
+
+    fn validate(&self) -> Result<TradeFilter, TradeFilterError> {
+        // Ensure that not both owner and order_uid are specified
+        match (self.order_uid.as_ref(), self.owner.as_ref()) {
+            (Some(_), Some(_)) => Err(TradeFilterError::InvalidFilter(
+                "Cannot specify both owner and order_uid".to_owned(),
+            )),
+            _ => Ok(self.trade_filter()),
+        }
```

Comment:
Maybe slightly cleaner is something like
```rust
if self.order_uid.is_some() && self.owner.is_some() {
    return Err(...)
}
Ok(...)
```

Reply:
Right yes, I was initially thinking this would leave room for other possibilities but this can easily be adapted in the future. Will use your suggestion.